### ISSUE TITLE
False positives for Replaced Elements (img, button etc) in declaration-block-no-ignored-properties

### ISSUE THREAD

**Issue by ryanheathers** [stylelint/stylelint#1786]
`button` elements respect `width` properties at all times, counter-intuitively even when the `button` is set to `display: inline`. See this [Stack Overflow](http://stackoverflow.com/questions/27764122/html-button-is-a-inline-element-but).

The issue is that Stylelint's `declaration-block-no-ignored-properties` throws a warning for `width` (including min & max variants) **for any selector** that also has a `display: inline` declaration in the same block.

This is causing false positive warnings for Button elements. (Note that it may be causing similar issues for certain other HTML elements that are treated as [Replaced Elements](https://developer.mozilla.org/en-US/docs/Web/CSS/Replaced_element) in CSS, although I have not tested this and do not know this for a fact.)

For example, this code will be flagged:

``` css
button {
  display: inline;
  max-width: 200px;
}
```

I looked at the rule's source for a possible easy fix, but didn't see an obvious one that would scale well if more elements than just `button` are affected by this CSS behavior. Possibly some kind of `exceptions` sub-array of HTML elements could be maintained for each `ignored` block but that may also require a lot of logic refactoring.

**Comment by evilebottnawi** [comment]
@ryanheathers you are right. Can you collect all replaced elements or find url with list their? I can add array of replaced element and ignore this as `type` selector, but it is forever be false position for this:
```css
.some-button {
  display: inline;
  max-width: 200px;
}
```
For above example i propose to add `ignoreSelector` (support string and regex). What do you think?

**Comment by kevinSuttle** [comment]
Here's one for you. 

```diff
.foo  {
  background-color: transparent;
  border: none;
-  display: inline;
  position: absolute;
  right: 0;
+  display: inline-block;
  padding: 0;
  height: 42px;
  width: 30px;
}
```

Even though that 2nd `display` declaration overrides the first, this stylelint rule still triggers because I set a `height` and `width`. 

What would be nice would be if stylelint saw the overridden declaration and warned me.

**Comment by jeddy3** [comment]
@kevinSuttle Can you open a separate issue for that please. It's definitely a short coming we should address. 

> I can add array of replaced element and ignore this as type selector

SGTM.

> but it is forever be false position for this: For above example i propose to add ignoreSelector (support string and regex). What do you think?

Should we just document this limitation and then rely on the `stylelint-disable` commands, rather than adding an `ignoreSelector` option? That is what the `stylelint-disable` commands are there for, and this has the advantage of keeping the ignore command together with the specific rule rather than sticking it in the config. The other `ignore*` options generally apply to more than one instance in a source. This feels like each ignore item will be for one instance.

**Comment by kevinSuttle** [comment]
@jeddy3 Done. https://github.com/stylelint/stylelint/issues/1790

**Comment by jeddy3** [comment]
I'm going to label as "help wanted". Lets just patch the `button` *type selector* behaviour and add caveat to the README saying use disable commands to work around other "Replaced Elements".

**Comment by jeddy3** [comment]
The [plan now](https://github.com/stylelint/stylelint/pull/1906#issuecomment-247869191) is to simply remove the `display: inline` with `width` check from the rule.

**Comment by jeddy3** [comment]
@davidtheclark I'll jump on this.

**Comment by jeddy3** [comment]
Scratch that sorry, I'll [leave this](https://github.com/stylelint/stylelint/pull/1906#issuecomment-247860688) to @evilebottnawi to experiment with. I'm unsure how it will be possible to ensure the accuracy of this rule as the CSS has no way of knowing if it's being applied to a Replaced Element or not. So, I'm also in favour moving this out of core.

**Comment by jeddy3** [comment]
Closing in favour of https://github.com/stylelint/stylelint/issues/2054
